### PR TITLE
426 cover pt edge cases

### PR DESCRIPTION
**Pull request by heshamMassoud**
#### Summary
Adds more tests to improve test coverage of product type sync.

#### Description
Adds more tests to improve test coverage of product type sync.

#### Relevant Issues
#426

### CODE REVIEW

File: src/test/java/com/commercetools/sync/producttypes/utils/producttypeactionutils/BuildAttributeDefinitionUpdateActionsTest.java [hunk 1]
```diff
@@ -881,6 +926,135 @@ void buildAttributesUpdateActions_WithSetOfIdenticalLEnums_ShouldBuildNoActions(
...
+            AddAttributeDefinition.of(AttributeDefinitionDraftBuilder
+                .of(newDefinition.getAttributeType(),
+                    newDefinition.getName(), newDefinition.getLabel(),
+                    newDefinition.isRequired())
+                .build())
+        );
+    }
+
+    @Test
+    void buildAttributesUpdateActions_WithOldLenumAndNewAsNonLenum_ShouldBuildRemoveAndAddActions() {
```
Comment (lojzatran) [inline src/test/java/com/commercetools/sync/producttypes/utils/producttypeactionutils/BuildAttributeDefinitionUpdateActionsTest.java]: What's the difference between this test and the other test `buildAttributesUpdateActions_WithNewLenumAndOldAsNonLenum_ShouldBuildRemoveAndAddActions` that you have to make separate tests?

File: src/integration-test/java/com/commercetools/sync/integration/externalsource/producttypes/ProductTypeWithNestedAttributeSyncIT.java [hunk 2]
```diff
@@ -245,6 +253,86 @@ void sync_WithNewProductTypeWithANonExistingReference_ShouldCreateProductType()
...
+            PRODUCT_TYPE_DESCRIPTION_4,
+            singletonList(ATTRIBUTE_DEFINITION_DRAFT_3));
+
+        final SphereClient ctpClient = spy(CTP_TARGET_CLIENT);
+        final BadGatewayException badGatewayException = new BadGatewayException();
+        when(ctpClient.execute(any(ProductTypeQuery.class)))
+            .thenCallRealMethod() // should work on caching
+            .thenCallRealMethod() // should work when fetching matching product types
+            .thenCallRealMethod() // should work when second fetching matching product types
+            .thenReturn(exceptionallyCompletedFuture(badGatewayException)) // fail on fetching during resolution
```
[the remark this answers is not in the document]
Comment (heshamMassoud) [reply src/integration-test/java/com/commercetools/sync/integration/externalsource/producttypes/ProductTypeWithNestedAttributeSyncIT.java]: a missing reference is not a "fail".. It will be counted as a missing reference and the sync would count it as missing but in this case we want to count it as a fail

File: src/integration-test/java/com/commercetools/sync/integration/externalsource/producttypes/ProductTypeWithNestedAttributeSyncIT.java [hunk 3]
```diff
@@ -245,6 +253,86 @@ void sync_WithNewProductTypeWithANonExistingReference_ShouldCreateProductType()
...
+    @Test
+    void sync_WithNewProductTypeWithFailedFetchOnReferenceResolution_ShouldFail() {
+        // preparation
+        final AttributeDefinitionDraft nestedTypeAttr = AttributeDefinitionDraftBuilder
+            .of(AttributeDefinitionBuilder
+                .of("nestedattr", ofEnglish("nestedattr"),
+                    NestedAttributeType.of(ProductType.referenceOfId(PRODUCT_TYPE_KEY_4)))
+                .build())
+            // "isSearchable=true is not supported for attribute type 'nested'."
+            .searchable(false)
```
[the remark this answers is not in the document]
Comment (heshamMassoud) [reply src/integration-test/java/com/commercetools/sync/integration/externalsource/producttypes/ProductTypeWithNestedAttributeSyncIT.java]: yes :) I thought it would be implied from the comment above.

File: src/integration-test/java/com/commercetools/sync/integration/externalsource/producttypes/ProductTypeWithNestedAttributeSyncIT.java [hunk 4]
```diff
@@ -245,6 +253,86 @@ void sync_WithNewProductTypeWithANonExistingReference_ShouldCreateProductType()
...
+            .build();
+
+        final ProductTypeDraft newProductTypeDraft = ProductTypeDraft.ofAttributeDefinitionDrafts(
```
[the remark this answers is not in the document]
Comment (heshamMassoud) [reply src/integration-test/java/com/commercetools/sync/integration/externalsource/producttypes/ProductTypeWithNestedAttributeSyncIT.java]: but it's not a wrong nested type :)?

File: src/integration-test/java/com/commercetools/sync/integration/externalsource/producttypes/ProductTypeWithNestedAttributeSyncIT.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (lojzatran) [reply src/integration-test/java/com/commercetools/sync/integration/externalsource/producttypes/ProductTypeWithNestedAttributeSyncIT.java]: Actually if you look in the source code, it implies that `isSearchable = null` by default

File: src/integration-test/java/com/commercetools/sync/integration/externalsource/producttypes/ProductTypeWithNestedAttributeSyncIT.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (lojzatran) [reply src/integration-test/java/com/commercetools/sync/integration/externalsource/producttypes/ProductTypeWithNestedAttributeSyncIT.java]: Then a nested type with missing reference? 😄

File: src/main/java/com/commercetools/sync/producttypes/ProductTypeSync.java [hunk 5]
```diff
@@ -226,20 +223,9 @@ private void handleError(@Nonnull final String errorMessage, @Nullable final Thr
             .stream()
             .map(newProductType ->
                 removeAndKeepTrackOfMissingNestedAttributes(newProductType, keyToIdCache))
-            .map(draftWithoutMissingRefAttrs ->
-                referenceResolver
-                    .resolveReferences(draftWithoutMissingRefAttrs)
-                    .thenCompose(resolvedDraft -> syncDraft(oldProductTypeMap, resolvedDraft))
-                    .exceptionally(completionException -> {
```
[the remark this answers is not in the document]
Comment (heshamMassoud) [reply src/main/java/com/commercetools/sync/producttypes/ProductTypeSync.java]: I removed the exceptionally part, because I don't expect any error to occur at this point. If any occur, then it must be programmatical error or a case that we didn't think of.

File: src/test/java/com/commercetools/sync/producttypes/ProductTypeSyncTest.java [hunk 6]
```diff
@@ -42,6 +43,153 @@
...
+
+        // assertions
+        assertThat(errorMessages).isEmpty();
+        assertThat(exceptions).isEmpty();
+
+        assertThat(productTypeSyncStatistics).hasValues(1, 0, 1, 0, 0);
+    }
+
+    @Test
+    void sync_WithNullAttributeDefinitions_ShouldSyncCorrectly() {
```
[the remark this answers is not in the document]
Comment (heshamMassoud) [reply src/test/java/com/commercetools/sync/producttypes/ProductTypeSyncTest.java]: I guess it won't hurt but it's not the purpose of the test.

File: src/test/java/com/commercetools/sync/producttypes/ProductTypeSyncTest.java [hunk 7]
```diff
@@ -42,6 +43,153 @@
...
+    }
+
+    @Test
+    void sync_WithNullInAttributeDefinitions_ShouldSyncCorrectly() {
+        // preparation
+        final ProductTypeDraft newProductTypeDraft = ProductTypeDraft.ofAttributeDefinitionDrafts(
+            "foo",
+            "name",
+            "desc",
+            singletonList(null)
```
[the remark this answers is not in the document]
Comment (heshamMassoud) [reply src/test/java/com/commercetools/sync/producttypes/ProductTypeSyncTest.java]: why not? you cannot force the user not to do it, as they create their own drafts as this line above.

[the remark this answers is not in the document]
Comment (lojzatran) [reply src/test/java/com/commercetools/sync/producttypes/ProductTypeSyncTest.java]: I mean if the platform even accept such input.

File: src/test/java/com/commercetools/sync/producttypes/utils/producttypeactionutils/BuildAttributeDefinitionUpdateActionsTest.java (hunk 1 above)
Comment (heshamMassoud) [reply src/test/java/com/commercetools/sync/producttypes/utils/producttypeactionutils/BuildAttributeDefinitionUpdateActionsTest.java]: This is the difference: `WithOldLenumAndNewAsNonLenum` and `WithNewLenumAndOldAsNonLenum` :)

It's trivial but they are 2 different cases.

File: src/test/java/com/commercetools/sync/producttypes/ProductTypeSyncTest.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (lojzatran) [reply src/test/java/com/commercetools/sync/producttypes/ProductTypeSyncTest.java]: But how do you know you sync correctly then?

File: src/test/java/com/commercetools/sync/producttypes/utils/producttypeactionutils/BuildAttributeDefinitionUpdateActionsTest.java (hunk 1 above)
Comment (lojzatran) [reply src/test/java/com/commercetools/sync/producttypes/utils/producttypeactionutils/BuildAttributeDefinitionUpdateActionsTest.java]: But why you test these cases and for example not case like `oldSetAndNewAsLenum` etc.? 😄

File: src/integration-test/java/com/commercetools/sync/integration/externalsource/producttypes/ProductTypeWithNestedAttributeSyncIT.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (heshamMassoud) [reply src/integration-test/java/com/commercetools/sync/integration/externalsource/producttypes/ProductTypeWithNestedAttributeSyncIT.java]: hmm probably you're looking at the wrong place. I will add more info: 050eef1

File: src/test/java/com/commercetools/sync/producttypes/ProductTypeSyncTest.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (heshamMassoud) [reply src/test/java/com/commercetools/sync/producttypes/ProductTypeSyncTest.java]: yes

File: src/test/java/com/commercetools/sync/producttypes/ProductTypeSyncTest.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (heshamMassoud) [reply src/test/java/com/commercetools/sync/producttypes/ProductTypeSyncTest.java]: that it didn't fail and got updated but I will add an assert for the actions now to be more strict.

[the remark this answers is not in the document]
Comment (heshamMassoud) [reply src/test/java/com/commercetools/sync/producttypes/ProductTypeSyncTest.java]: 050eef1

File: src/test/java/com/commercetools/sync/producttypes/utils/producttypeactionutils/BuildAttributeDefinitionUpdateActionsTest.java (hunk 1 above)
Comment (heshamMassoud) [reply src/test/java/com/commercetools/sync/producttypes/utils/producttypeactionutils/BuildAttributeDefinitionUpdateActionsTest.java]: but this is tested here: `buildAttributesUpdateActions_WithSetOfText_ShouldBuildActions` (implicitly)

File: src/test/java/com/commercetools/sync/producttypes/ProductTypeSyncTest.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (lojzatran) [reply src/test/java/com/commercetools/sync/producttypes/ProductTypeSyncTest.java]: https://github.com/commercetools/commercetools-sync-java/pull/462/commits/7fbde55414f47d7e03c971fad53dfa408c3e25d4

File: src/integration-test/java/com/commercetools/sync/integration/externalsource/producttypes/ProductTypeWithNestedAttributeSyncIT.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (heshamMassoud) [reply src/integration-test/java/com/commercetools/sync/integration/externalsource/producttypes/ProductTypeWithNestedAttributeSyncIT.java]: 696e536
